expat/tests/runtests.c

```diff
@@ -11975,6 +12022,209 @@ START_TEST(test_nsalloc_prefixed_element)
+        "  <!ELEMENT he (#PCDATA)*>\n"
+        "  <!ENTITY large '"
+        /* 64 character on each line */
+        /* ...gives a total length of 1024 */
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "'>\n"
+        "]>\n"
+        "<he>&large;</he>\n";
+    expect_failure(text, XML_ERROR_ENTITY_VIOLATION_SIZE,
+                   "XML_ERROR_ENTITY_VIOLATION_SIZE not raised");
+}
+END_TEST
+
+START_TEST(test_huge_xml_entity_expansion_limit)
+{
+    const char *text =
+        "<!DOCTYPE he ["
+        "<!ELEMENT he (#PCDATA)*>"
+        "<!ENTITY e '"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "'>]><he>"
+        /* 50 + 64 + 8 + (36 * 3) == 230
+         * 36 expansions of 64 bytes == 2304 */
+        "&e;&e;&e;&e;&e;&e;&e;&e;&e;&e;"
+        "&e;&e;&e;&e;&e;&e;&e;&e;&e;&e;"
+        "&e;&e;&e;&e;&e;&e;&e;&e;&e;&e;"
+        "&e;&e;&e;&e;&e;&e;"
+        "</he>\n";
+    expect_failure(text, XML_ERROR_ENTITY_VIOLATION_RATIO,
+                   "XML_ERROR_ENTITY_VIOLATION_RATIO not raised");
+}
+END_TEST
+
+START_TEST(test_huge_xml_hash_table_limit)
+{
+    XML_Size tableEntries = 4; /* smaller than 4 entities + 1 element */
+    const char *text =
+        "<!DOCTYPE he [\n"
+        "  <!ELEMENT he (#PCDATA)*>\n"
+        "  <!ENTITY e1 '&e2;'>\n"
+        "  <!ENTITY e2 '&e3;'>\n"
+        "  <!ENTITY e3 '&e4;'>\n"
+        "  <!ENTITY e4 'entity'>\n"
+        "]>\n"
+        "<he>&e1;</he>\n";
+
+    if (XML_SetOption(parser, XML_OPTION_MAX_HASH_TABLE_ENTRIES, &tableEntries) != XML_STATUS_OK)
+        fail("XML_SetOptions(XML_OPTION_MAX_HASH_TABLE_ENTRIES) failed");
+
+    expect_failure(text, XML_ERROR_HASH_TABLE_VIOLATION,
+                   "XML_ERROR_HASH_TABLE_VIOLATION not raised");
+}
+END_TEST
+
+START_TEST(test_huge_xml_enabled)
```

Comment:
The purpose of this test seems to verify that `XML_OPTION_HUGE_XML = XML_TRUE` makes the parser bypass limit code, no matter how tiny set limits might be. If that's true, I'd propose to rename the test to something like `test_huge_xml_bypasses_all_limits` to make things more obvious. Thanks!